File: Computer_Science/Algorithm/Learning/Algorithms/双指针/滑动窗口/模板题/code/3.无重复字符的最长子串.cpp
/*
 * @lc app=leetcode.cn id=3 lang=cpp
 *
 * [3] 无重复字符的最长子串
 */

#include <unordered_set>
#include <string>
using namespace std;
// @lc code=start
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.length();
        int left = 0;
        int ans = 0;
        unordered_set<char> hash;
        for (int right = 0; right < n; right++) {
            char c = s[right];
            while (hash.count(c) > 0){
                hash.erase(s[left]);
                left++;
            }
            hash.insert(c);
            ans = max(ans, right - left + 1);
        }

        return ans;
    }
};
// @lc code=end

File: 4work/Algorithm/leetcode/hot100/3.无重复字符的最长子串.java
/*
 * @lc app=leetcode.cn id=3 lang=java
 *
 * [3] 无重复字符的最长子串
 */

// @lc code=start

import java.util.HashSet;
import java.util.Set;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        Set<Character> set = new HashSet<>();
        char[] arr = s.toCharArray();
        int left = 0;
        int ans = 0;
        int len = 0;
        for (int right = 0; right < arr.length; right++) {
            char ch = arr[right];
            // 如果存在, 则向右移动窗口左端点, 直到弹出ch元素
            while (set.contains(ch)) {
                set.remove(arr[left++]);
                len--;
            }
            set.add(ch);
            len++;
            ans = Math.max(len, ans);
        }
        return ans;
    }
}
// "pwwkew"\n
// @lc code=end

